modules/bmc/basescanner.rb

```diff
@@ -0,0 +1,50 @@
+require 'ipaddr'
+
+module Proxy
+  module BMC
+    # This is the interface for scanning BMC IP address ranges.
+    class BaseScanner
+      def initialize(args = { })
+        if args.key? :address_first
+          address_first = IPAddr.new args[:address_first]
+          address_last  = IPAddr.new args[:address_last]
+          @range = (address_first..address_last)
+        elsif args.key?(:address) && args.key?(:netmask)
+          @range = IPAddr.new("#{args[:address]}/#{args[:netmask]}").to_range
+        else
+          @range = IPAddr.new(args[:address]).to_range
+        end
+        # Disallow range too large
+        scanner_max_range_size = max_range_size
+        if @range.first(scanner_max_range_size+1).size > scanner_max_range_size
+          @range = nil
+          @invalid_reason = "Range too large. Batch supports only #{scanner_max_range_size} IP addresses at a time."
+        end
+      rescue
+        @range = nil
+        if args.is_a?(Hash) && args.key?(:address)
+          @invalid_reason = "Invalid CIDR provided"
+        else
+          @invalid_reason = "Invalid IP address provided"
+        end
+      end
+
+      def valid?
+        @range.is_a? Range
```

Comment:
Instead of validating ip range size in constructor, I would suggest saving ip range parameters there, and delay the actual range initialization until later (you are doing an explicit validity check anyway).

For example, something like this might work:
```
def range
  @range ||= (@address_first..@address_last)
  raise "Range too large." if @range.first(scanner_max_range_size+1).size > scanner_max_range_size
end
```
This will allow you to use exceptions too.